README.md
# heatmap-scatter-dash

A heatmap-scatterplot using [Dash by plotly](https://plot.ly/products/dash/).
From the commandline it can be started on localhost or AWS, or it can be run
from the [Refinery](https://github.com/refinery-platform/refinery-platform) GUI.

![v0.1.2-screenshot](https://user-images.githubusercontent.com/730388/36332441-2ba33e9c-1340-11e8-900a-6f16549f1f6b.png)

```
$ python app_runner.py -h
usage: app_runner.py [-h] (--demo ROWS COLS META | --files CSV [CSV ...])
                     [--diffs CSV [CSV ...]] [--metas CSV [CSV ...]]
                     [--most_variable_rows ROWS] [--html_table]
                     [--truncate_table N] [--port PORT]
                     [--p_value_re RE [RE ...]] [--log_fold_re RE [RE ...]]
                     [--profile [DIR]] [--html_error] [--debug]
                     [--api_prefix PREFIX]

Light-weight visualization for differential expression

optional arguments:
  -h, --help            show this help message and exit
  --demo ROWS COLS META
                        Generates a random matrix with the number of rows and
                        columns specified. In addition, "META" determines the
                        number of mock metadata fields to associate with each
                        column.
  --files CSV [CSV ...]
                        Read CSV or TSV files. Identifiers should be in the
                        first column and multiple files will be joined on
                        identifier. Gzip files are also handled.
  --diffs CSV [CSV ...]
                        Read CSV or TSV files containing differential
                        expression data.
  --metas CSV [CSV ...]
                        Read CSV or TSV files containing metadata: Row labels
                        should match column headers of the raw data.
  --most_variable_rows ROWS
                        For the heatmap, we first sort by row variance, and
                        then take the number of rows specified here. Defaults
                        to 500.
  --html_table          The default is to use pre-formatted text for the
                        tables. HTML tables are available, but are twice as
                        slow.
  --truncate_table N    Truncate the table to the first N rows. Table
                        rendering is often a bottleneck. Default is not to
                        truncate.
  --port PORT           Specify a port to run the server on. Defaults to 8050.

Refinery/Developer:
  These parameters will probably only be of interest to developers, and/or
  they are used when the tool is embedded in Refinery.

  --p_value_re RE [RE ...]
                        Regular expressions which column headers will be
                        checked against to identify p-values. Defaults to
                        ['p.*value', 'padj', 'fdr'].
  --log_fold_re RE [RE ...]
                        Regular expressions which column headers will be
                        checked against to identify fold-change values.
                        Defaults to ['\\blog[^a-z]'].
  --profile [DIR]       Saves a profile for each request in the specified
                        directory, "/tmp" by default. Profiles can be viewed
                        with snakeviz.
  --html_error          If there is a configuration error, instead of exiting,
                        start the server and display an error page.
  --debug               Run the server in debug mode: The server will restart
                        in response to any code changes, and some hidden
                        fields will be shown.
  --api_prefix PREFIX   Prefix for API URLs.
```

## Getting Started

### Docker

If you have Docker installed, and data available at public URLs,
this is the easiest way to get started:

```bash
$ V=v0.1.5
$ FIXTURES=https://raw.githubusercontent.com/refinery-platform/heatmap-scatter-dash/$V/fixtures/good/data
$ docker run --name heatmap --detach --publish 8888:80 \
  -e "FILE_URLS=$FIXTURES/counts.csv $FIXTURES/counts-copy.csv.gz" \
  mccalluc/heatmap_scatter_dash:$V
```

Then visit [http://localhost:8888](http://localhost:8888).

If multiple URLs are desired, use spaces in the value
of the environment variables, as in the example.
Besides providing count data, you can also specify
`DIFF_URLS` for differential expression data and
`META_URLS` for metadata.


### From Source

Check out the project and install dependencies:
```bash
  # Requires Python3:
$ python --version
$ git clone https://github.com/refinery-project/heatmap-scatter-dash.git
$ cd heatmap-scatter-dash
$ pip install -r requirements-freeze.txt
```

Then run it locally:

```bash
$ cd context

  # Generate a random matrix:
$ ./app_runner.py --demo 100 10 5

  # Load data from disk:
$ ./app_runner.py --files ../fixtures/good/data/counts.csv \
                  --diffs ../fixtures/good/data/stats-* \
                  --meta ../fixtures/good/data/metadata.csv
```

and visit `http://localhost:8050/`.

## Testing

One bash script, `test.sh`, handles all our tests:
- Python unit tests
- Python style tests (`flake8` and `isort`)
- Cypress.io interaction tests
- Docker container build and launch

A few more dependencies are required for this to work locally:
```bash
  # Install Docker: https://www.docker.com/community-edition
$ pip install -r requirements-dev.txt
$ npm install cypress --save-dev
```

## Release

Successful Github tags and PRs will prompt Travis to push the built image to Dockerhub. For a new version number:

```bash
$ git tag v0.0.x && git push origin --tags
```

## For More Information...

There are a few [notes](docs) on implementation decisions and lessons learned.

The [online help](context/app/help/help.md) can be previewed to get a better sense of the operational details.


CHANGELOG.md
# Change Log

## [v0.1.15](https://github.com/refinery-platform/heatmap-scatter-dash/tree/v0.1.15) (Aug 13, 2018)

* Fix race condition where selecting points on the scatterplot did not always
update the found set.
* Also look for "padj" and "fdr" when scanning headers.
* Move help to tab instead of separate page.


## [v0.1.14](https://github.com/refinery-platform/heatmap-scatter-dash/tree/v0.1.14) (Aug 2, 2018)

* app_runner_refinery sniffs files to determine if they are counts or differential expression.


## [v0.1.13](https://github.com/refinery-platform/heatmap-scatter-dash/tree/v0.1.13) (Jul 26, 2018)

* If the body of the table has non-numeric columns, the value in the first
of those columns is prepended to the row label. (Often, the first column is
just a numeric ID, but something more human-readable is present farther to
the right.)


## [v0.1.12](https://github.com/refinery-platform/heatmap-scatter-dash/tree/v0.1.12) (Jul 10, 2018)

* Support inputs with just two columns and numeric gene IDs.


## [v0.1.11](https://github.com/refinery-platform/heatmap-scatter-dash/tree/v0.1.11) (Jun 13, 2018)

* By default, drop non-numeric values from dataframe.


## [v0.1.10](https://github.com/refinery-platform/heatmap-scatter-dash/tree/v0.1.10) (May 2, 2018)

* Big speedup by using preformatted output rather than html for table.


## [v0.1.9](https://github.com/refinery-platform/heatmap-scatter-dash/tree/v0.1.9) (Apr 30, 2018)

* Support GCT files.


context/app_runner.py
#!/usr/bin/env python
import argparse
import html
import traceback
from os import makedirs
from os.path import abspath, basename
from warnings import warn

import numpy as np
import pandas
from dataframer import dataframer
from flask import Flask, send_from_directory
from werkzeug.contrib.profiler import ProfilerMiddleware

from app.download.download_app import DownloadApp
from app.utils.frames import find_index, merge
from app.utils.vulcanize import Vulcanizer
from app.vis.callbacks import VisCallbacks

P_VALUE_RE_LIST = [r'p.*value', r'padj', r'fdr']
LOG_FOLD_RE_LIST = [r'\blog[^a-z]']


def file_dataframes(files, relabel=False):
    frames = []
    label_maps = []
    for file in files:
        df_info = dataframer.parse(file, relabel=relabel)
        df = df_info.data_frame
        df.index = [str(i) for i in df.index]
        frames.append(df)
        label_maps.append(df_info.label_map)
        file.close()
    return (frames, label_maps)


def demo_dataframes(rows, cols, metas):
    data_array = np.random.rand(rows, cols)
    cond_labels = ['cond-{}'.format(i) for i in range(cols)]
    gene_labels = ['gene-{}'.format(i) for i in range(rows)]

    meta_array = np.random.rand(cols, metas)
    meta_labels = ['meta-{}'.format(i) for i in range(metas)]
    return (
        [pandas.DataFrame(data_array, columns=cond_labels, index=gene_labels)],
        [pandas.DataFrame(meta_array, columns=meta_labels, index=cond_labels)]
    )


def init(args, parser):  # TODO: Why is parser here?
    if args.files:
        (dataframes, label_maps) = file_dataframes(args.files, relabel=True)
        (meta_dataframes, _) = file_dataframes(args.metas)
    elif args.demo:
        (dataframes, meta_dataframes) = demo_dataframes(*args.demo)
        label_maps = []
    else:
        # Argparser validation should keep us from reaching this point.
        raise Exception('Either "demo" or "files" is required')

    union_dataframe = merge(dataframes)
    union_label_map = {}
    for lm in label_maps:
        union_label_map.update(lm)
    genes = set(union_dataframe.index.tolist())
    if args.diffs:
        diff_dataframes = {}
        for diff_file in args.diffs:
            df_info = dataframer.parse(
                diff_file,
                col_zero_index=False,
                keep_strings=True
            )
            key = basename(diff_file.name)
            value = Vulcanizer(
                log_fold_re_list=[r'\blog[^a-z]'],
                p_value_re_list=[r'p.*value']
            ).vulcanize(find_index(df_info.data_frame, genes))
            diff_dataframes[key] = value
    else:
        diff_dataframes = {
            'No differential files given': pandas.DataFrame()
        }

    union_meta_dataframe = merge(meta_dataframes)

    server = Flask(__name__, static_url_path='')

    server.route('/static/<path:path>')(
        lambda path: send_from_directory('app/static', path)
    )

    # TODO: Just calling constructor shouldn't do stuff.
    VisCallbacks(
        server=server,
        url_base_pathname='/',
        union_dataframe=union_dataframe,
        union_label_map=union_label_map,
        diff_dataframes=diff_dataframes,
        meta_dataframe=union_meta_dataframe,
        api_prefix=args.api_prefix,
        debug=args.debug,
        most_variable_rows=args.most_variable_rows,
        html_table=args.html_table,
        truncate_table=args.truncate_table
    )
    DownloadApp(
        server=server,
        url_base_pathname='/download',
        dataframe=union_dataframe
    )
    return server


def main(args, parser=None):
    try:
        app = init(args=args, parser=parser)
        if args.profile:
            abs_profile_path = abspath(args.profile)
            makedirs(abs_profile_path, exist_ok=True)
            app.wsgi_app = ProfilerMiddleware(app.wsgi_app,
                                              profile_dir=abs_profile_path)
        app.run(
            debug=args.debug,
            port=args.port,
            host='0.0.0.0'
        )
    except Exception as e:
        # Big try-blocks are usually to be avoided...
        # but here, we want to be sure that some server comes up
        # and returns a 200, so that django-proxy knows to stop waiting,
        # and the end-user can see the error.
        if not args.html_error:
            raise Exception(e)
        app = Flask('error-page')

        args_str = str(args)
        warn(args_str)

        error_str = ''.join(
            traceback.TracebackException.from_exception(e).format()
        )
        warn(error_str)

        @app.route("/")
        def error_page():
            return '''
                <html><head><title>Error</title></head><body>
                <p>An error has occurred: There may be a problem with the
                input provided. To report a bug, please copy this page and
                paste it in a <a href="{url}">bug report</a>. Thanks!</p>
                <pre>{args}</pre>
                <pre>{stack}</pre>
                </body></html>'''.format(
                url='https://github.com/refinery-platform/'
                'heatmap-scatter-dash/issues',
                args=html.escape(args_str),
                stack=html.escape(error_str))

        app.run(
            port=args.port,
            host='0.0.0.0'
        )


def arg_parser():
    parser = argparse.ArgumentParser(
        description='Light-weight visualization for differential expression')

    binary_file = argparse.FileType('rb')
    input_source = parser.add_mutually_exclusive_group(required=True)
    input_source.add_argument(
        '--demo', nargs=3, type=int, metavar=('ROWS', 'COLS', 'META'),
        help='Generates a random matrix with the number of rows and columns '
             'specified. In addition, "META" determines the number of mock '
             'metadata fields to associate with each column.')
    input_source.add_argument(
        '--files', nargs='+', metavar='CSV', type=binary_file,
        help='Read CSV or TSV files. Identifiers should be in the first '
             'column and multiple files will be joined on identifier. '
             'Gzip files are also handled.')

    parser.add_argument(
        '--diffs', nargs='+', metavar='CSV',
        type=binary_file, default=(),
        help='Read CSV or TSV files containing differential expression data.')

    parser.add_argument(
        '--metas', nargs='+', metavar='CSV',
        type=binary_file, default=(),
        help='Read CSV or TSV files containing metadata: Row labels should '
             'match column headers of the raw data.')

    parser.add_argument(
        '--most_variable_rows', type=int, default=500, metavar='ROWS',
        help='For the heatmap, we first sort by row variance, and then take '
             'the number of rows specified here. Defaults to 500.')

    parser.add_argument(
        '--html_table', action='store_true',
        help='The default is to use pre-formatted text for the tables. '
             'HTML tables are available, but are twice as slow.')

    parser.add_argument(
        '--truncate_table', type=int, default=None, metavar='N',
        help='Truncate the table to the first N rows. Table rendering is '
             'often a bottleneck. Default is not to truncate.')

    parser.add_argument(
        '--port', type=int, default=8050,
        help='Specify a port to run the server on. Defaults to 8050.')

    group = parser.add_argument_group(
        'Refinery/Developer',
        'These parameters will probably only be of interest to developers, '
        'and/or they are used when the tool is embedded in Refinery.')

    group.add_argument(
        '--p_value_re', nargs='+', type=str, metavar='RE',
        default=P_VALUE_RE_LIST,
        help='Regular expressions which column headers will be checked '
             'against to identify p-values. Defaults to {}.'.format(
                 repr(P_VALUE_RE_LIST)
             ))
    group.add_argument(
        '--log_fold_re', nargs='+', type=str, metavar='RE',
        default=LOG_FOLD_RE_LIST,
        help='Regular expressions which column headers will be checked '
             'against to identify fold-change values. Defaults to {}.'.format(
                 repr(LOG_FOLD_RE_LIST)
             ))
    group.add_argument(
        '--profile', nargs='?', type=str, default='/tmp', metavar='DIR',
        help='Saves a profile for each request in the specified directory, '
             '"/tmp" by default. Profiles can be viewed with snakeviz.')
    group.add_argument(
        '--html_error', action='store_true',
        help='If there is a configuration error, instead of exiting, '
             'start the server and display an error page.')
    group.add_argument(
        '--debug', action='store_true',
        help='Run the server in debug mode: The server will '
             'restart in response to any code changes, '
             'and some hidden fields will be shown.')
    group.add_argument(
        '--api_prefix', default='', metavar='PREFIX',
        help='Prefix for API URLs.')

    return parser


if __name__ == '__main__':
    parser = arg_parser()
    args = parser.parse_args()
    main(args, parser)
